blogposts/2020/evolution-of-the-precise-code-intel-backend.md

```diff
@@ -0,0 +1,175 @@
+
+Well, shoot.
+
+Our path forward was to add [another level of indirection](https://en.wikipedia.org/wiki/Indirection), which is honestly the solution _most_ problems. This change introduces a new lsif-dump-manager, known today as the precise-code-intel-bundle-manager, service that owns the persistent disk. 
+
+![architecture diagram](https://storage.googleapis.com/sourcegraph-assets/lsif-arch-7.png)
+
+This frees any responsibility of reading or writing to the disk from the lsif-server and the lsif-worker, making them free to scale horizontally and statelessly. Scaling the dump managers themselves is a different problem, but one we have some experience with already due to the sharded nature of our gitservers.
+
+## Rewrite services in Go
+
+**PRs**:
+- [Rewrite precise-code-intel-api-server in Go (#9703)](https://github.com/sourcegraph/sourcegraph/pull/9703/)
+- [Rewrite precise-code-intel-bundle-manager in Go (#9586)](https://github.com/sourcegraph/sourcegraph/pull/9586/)
+- [Rewrite precise-code-intel-worker in Go (#10105)](https://github.com/sourcegraph/sourcegraph/pull/10105)
+
+At this point, the code nav team had been split into the web and code intel for some time now. After a few changes in the code intel team's membership, we no longer had the same core competencies: we lost one fan of TypeScript and gained two fans of Go.
+
+In an effort to unify the precise code intel backends with the rest of the application, we decided to rewrite the services, slowly, into Go. Our high-level plan was to first extract the CPU-bound work done by the lsif-server into a high-performance Go process which could be called by the existing worker code (this is similar to the [strangler fig](https://martinfowler.com/bliki/StranglerFigApplication.html) pattern of refactoring). This part of the stack seemed to benefit the most from a rewrite, and allowed us to use existing knowledge to aggressively optimize Go rather than having to learn low-level optimization techniques for Node.js environments. Eventually, more and more of the worker would fall into the Go code's event horizon, and eventually the entire worker would be replaced.
+
+But actions don't always occur as they're planned.
+
+Turns out it's pretty easy to write the same system in a language you know well after banging out all the bugs in a different language, and I just rewrote all three services in a single pass. The resulting code wasn't particularly idiomatic as it really was meant to be a fast, behaviorally-equivalent translation into the target language. Constant, small refactorings have been occurring since the rewrite in order to shape the code into something more appropriate for the new environment.
```

Comment:
```suggestion
Turns out it's pretty easy to write the same system in a language you know well after fixing all the bugs in a different language, and I just rewrote all three services in a single pass. The resulting code wasn't particularly idiomatic as it really was meant to be a fast, behaviorally-equivalent translation into the target language. Constant, small refactorings have been occurring since the rewrite in order to shape the code into something more appropriate for the new environment.
```